test/jdk/java/util/concurrent/Executors/AutoShutdown.java

```diff
@@ -23,89 +23,114 @@
-        }};
-        for (Executor executor : executors)
-            executor.execute(task);
-        await(threadStarted);
-        pleaseProceed.countDown();
-        Arrays.fill(executors, null);   // make executors unreachable
-        boolean done = false;
-        for (long timeout = 1L; !done && timeout <= 128L; timeout *= 2) {
-            System.gc();
-            done = true;
-            for (WeakReference<Thread> ref : poolThreads) {
-                Thread thread = ref.get();
-                if (thread != null) {
-                    TimeUnit.SECONDS.timedJoin(thread, timeout);
-                    if (thread.isAlive())
-                        done = false;
-                }
-            }
+    /**
+     * SingleThreadExecutor with no worker threads.
+     */
+    @ParameterizedTest
+    @MethodSource("executors")
+    void testNoWorker(Supplier<ExecutorService> supplier) throws Exception {
+        ExecutorService executor = supplier.get();
+        ExecutorService delegate = getDelegate(executor);
+        executor = null;
+        gcAndAwaitTermination(delegate);
+    }
+
+    /**
+     * SingleThreadExecutor with an idle worker thread.
+     */
+    @ParameterizedTest
+    @MethodSource("executors")
+    void testIdleWorker(Supplier<ExecutorService> supplier) throws Exception {
+        ExecutorService executor = supplier.get();
+        // submit a task to get a worker to start
+        executor.submit(() -> null).get();
+        ExecutorService delegate = getDelegate(executor);
+        executor = null;
+        gcAndAwaitTermination(delegate);
+    }
+
+    /**
+     * SingleThreadExecutor with an active worker and queued tasks.
+     */
+    @ParameterizedTest
+    @MethodSource("executorAndQueuedTaskCounts")
+    void testActiveWorker(Supplier<ExecutorService> supplier,int queuedTaskCount) throws Exception {
+        ExecutorService executor = supplier.get();
+        // the worker will execute one task, the other tasks will be queued
+        int ntasks = 1 + queuedTaskCount;
+        AtomicInteger completedTaskCount = new AtomicInteger();
+        for (int i = 0; i < ntasks; i++) {
+            executor.submit(() -> {
+                Thread.sleep(Duration.ofMillis(500));
+                completedTaskCount.incrementAndGet();
+                return null;
+            });
         }
-        if (!done)
-            throw new AssertionError("pool threads did not terminate");
+        ExecutorService delegate = getDelegate(executor);
+        executor = null;
+        gcAndAwaitTermination(delegate);
+        assertEquals(ntasks, completedTaskCount.get());
+    }
+
+    /**
+     * Returns the delegate for the given ExecutorService. The given ExectorService
+     * must be a Executors$DelegatedExecutorService.
+     */
+    private ExecutorService getDelegate(ExecutorService executor) throws Exception {
+        Field eField = Class.forName("java.util.concurrent.Executors$DelegatedExecutorService")
```

Comment:
I was under the impression that VarHandle/MethodHandle are generally preferred for writing white box tests (as I did myself in various WhiteBox.java tests).

Reply:
I don't think so. If you search the tests for uses of "+open" then you'll see a variety of usages.